### PR TITLE
lazy decode messages

### PR DESCRIPTION
**Pull request by DeveloperInProgress**

### CODE REVIEW

**Comment by stwiname** [comment]
It would be good to do the same for decoding events and transactions where possible and have some tests that show that the lazy loading works where possible. We would have to use jest spy features.

File: packages/node/src/utils/cosmos.ts [hunk 1]
```diff
@@ -169,7 +192,9 @@ function wrapCosmosMsg(
     block: block,
     msg: {
       typeUrl: rawMessage.typeUrl,
-      ...api.decodeMsg<any>(rawMessage),
+      get decodedMsg() {
```
Comment (stwiname) [inline packages/node/src/utils/cosmos.ts:195]: Is this changing the CosmosMessage type? Will this require updating other projects?

File: packages/node/src/utils/cosmos.ts [hunk 2]
```diff
@@ -105,9 +118,19 @@ export function filterEvents(
...
+  const filteredEvents = events
+    .filter((event) => {
+      filters.find((filter) => filterEvent(event, filter));
+    })
+    .map((evt) => {
+      evt.msg.msg = {
+        ...evt.msg.msg,
+        ...evt.msg.msg.decodedMsg,
+      };
+      delete evt.msg.msg.decodedMsg;
```
Comment (stwiname) [inline packages/node/src/utils/cosmos.ts:130]: Is this hoisting the change from the wrappers? I think this is quite hard to follow. 

Would we be better off making a breaking change to get more usable types?

File: packages/node/src/utils/cosmos.ts [hunk 3]
```diff
@@ -64,9 +67,19 @@ export function filterMessages(
...
     filterOrFilters instanceof Array ? filterOrFilters : [filterOrFilters];
 
-  const filteredMessages = messages.filter((message) => {
-    filters.find((filter) => filterMessageData(message, filter));
-  });
+  const filteredMessages = messages
+    .filter((message) => {
+      filters.find((filter) => filterMessageData(message, filter));
+    })
+    .map((msg) => {
```
Comment (stwiname) [inline packages/node/src/utils/cosmos.ts:74]: Would this need to be done if there are no filters too?

Comment (DeveloperInProgress) [reply packages/node/src/utils/cosmos.ts:74]: I just realized that this function will be only be called for custom datasources. It's a seperate mechanism for runtime datasources: https://github.com/subquery/subql-cosmos/blob/3b4468f26316378767954b7b03508c986c3f3ee5/packages/node/src/indexer/indexer.manager.ts#L528

File: packages/node/src/utils/cosmos.ts (hunk 1 above)
Comment (DeveloperInProgress) [reply packages/node/src/utils/cosmos.ts:195]: Looks like that the data were originally accessed through msg property, will have to be accessed through msg.decodedMsg

File: packages/node/src/utils/cosmos.ts (hunk 3 above)
Comment (DeveloperInProgress) [reply packages/node/src/utils/cosmos.ts:74]: And it wouldn't be done if there are no filters but, it has to be done.

Comment (DeveloperInProgress) [reply packages/node/src/utils/cosmos.ts:74]: And this transformation doesn't happen for runtime datasources. `msg.decodedMsg` would still exist.

File: packages/node/src/utils/cosmos.ts (hunk 1 above)
Comment (DeveloperInProgress) [reply packages/node/src/utils/cosmos.ts:195]: We could look into a way of transforming it so that the projects or dictionary remains unaffected

File: packages/types/src/interfaces.ts [hunk 4]
```diff
@@ -39,7 +39,10 @@ export interface CosmosMessage {
   idx: number;
   block: CosmosBlock;
   tx: CosmosTransaction;
-  msg: any;
+  msg: {
+    typeUrl: string;
+    decodedMsg: any;
```
Comment (stwiname) [inline packages/types/src/interfaces.ts:44]: Can we make this generic? `CosmosMessage<T = any>` and `decodedMsg: T`

**Comment by stwiname** [comment]
@DeveloperInProgress I think this is looking good but I think wrapping the data types will cause everything to be loaded.

Problem is here: https://github.com/subquery/subql-cosmos/blob/main/packages/node/src/utils/cosmos.ts#L245-L255

I think we might need to lazy load the events/messages/transactions, that way if there's no event handlers then we don't wrap them.

Can we have a test that calls `fetchBlocksBatches` please.
